manticore/core/cpu/binja.py

```diff
@@ -0,0 +1,1188 @@
+import ctypes
+from functools import wraps
+import logging
+import re
+from collections import defaultdict
+
+from .x86 import AMD64Operand
+
+from .abstractcpu import Cpu, RegisterFile, Operand, Syscall
+from .cpufactory import CpuFactory
+from ...core.cpu.disasm import BinjaILDisasm
+from ..smtlib import Operators, BitVecConstant, operator
+from ...utils.helpers import issymbolic
+
+logger = logging.getLogger("CPU")
+
+
+class BinjaRegisterFile(RegisterFile):
+
+    def __init__(self, arch, platform_regs):
+        from binaryninja import Architecture
+        self.arch = arch
+        arch = Architecture[arch]
+        self.reg_aliases = {
+            'STACK' : str(arch.stack_pointer),
+            'PC' : self.get_pc(),
+            'CS' : 'cs',
+            'SS' : 'ss',
+            'DS' : 'ds',
+            'ES' : 'es'
+        }
+        super(BinjaRegisterFile, self).__init__(aliases=self.reg_aliases)
+
+        # Get "dummy" registers for flags. We append 'F' for consistency
+        # with the X86 CPU
+        f_regs = [f + "f" for f in arch.flags]
+        # get (full width) architecture registers
+        arch_regs = list(set([arch.regs[r].full_width_reg
+                              for r in arch.regs.keys()]))
+
+        self.pl2b_map, self.b2pl_map = binja_platform_regmap(arch_regs,
+                                                             f_regs,
+                                                             self.reg_aliases,
+                                                             platform_regs)
+
+        # FIXME FIX tuples
+        new_aliases = filter(lambda x: (x[0] not in self.reg_aliases.keys() and
+                                        x[1] is not None and
+                                        not isinstance(x[1], tuple)),
+                             self.pl2b_map.items())
+        self.reg_aliases.update(new_aliases)
+
+        # all regs are: architecture, aliases and flags
+        all_regs = arch_regs + self.reg_aliases.values() + f_regs
+        self.registers = {reg : 0 for reg in all_regs}
+
+        # FIXME get these from the platform!! they are already part of registers
+        self.segment_registers = (['cs', 'ds', 'es', 'ss'] +
+                                  ['fs', 'gs', 'fsbase', 'gsbase'])
+
+    def write(self, reg, value):
+        from binaryninja.enums import ImplicitRegisterExtend
+        from binaryninja import Architecture
+
+        arch = Architecture[self.arch]
+        r = self._alias(str(reg))
+        # if this is a custom register just write to the dictionary
+        if r not in arch.regs:
+            self.registers[r] = value
+            return
+
+        # get the ILRegister object from Architecture
+        ilreg = arch.regs.get(r, None)
+        if ilreg is None:
+            raise NotImplementedError
```

Comment:
Where is this caught? Maybe just do `arch.regs[r]` and let the `KeyError` propagate?